curvefs/src/metaserver/metaserver_service.cpp

```diff
@@ -0,0 +1,317 @@
+    uint32_t uid = request->uid();
+    uint32_t gid = request->gid();
+    uint32_t mode = request->mode();
+    FsFileType type = request->type();
+    std::string symlink;
+    if (type == FsFileType::TYPE_SYM_LINK) {
+        if (!request->has_symlink()) {
+            response->set_statuscode(MetaStatusCode::SYM_LINK_EMPTY);
+            return;
+        }
+
+        symlink = request->symlink();
+        if (symlink.empty()) {
+            response->set_statuscode(MetaStatusCode::SYM_LINK_EMPTY);
+            return;
+        }
+    }
+
+    MetaStatusCode status = inodeManager_->CreateInode(fsId, length, uid, gid,
+                            mode, type, symlink, response->mutable_inode());
+    response->set_statuscode(status);
+    if (status != MetaStatusCode::OK) {
+        response->clear_inode();
+    }
+    return;
+}
+
+void MetaServerServiceImpl::CreateRootInode(
+                   ::google::protobuf::RpcController* controller,
+                   const ::curvefs::metaserver::CreateRootInodeRequest* request,
+                   ::curvefs::metaserver::CreateRootInodeResponse* response,
+                   ::google::protobuf::Closure* done) {
+    brpc::ClosureGuard doneGuard(done);
+    brpc::Controller* cntl = static_cast<brpc::Controller*>(controller);
+    uint32_t fsId = request->fsid();
+    uint32_t uid = request->uid();
+    uint32_t gid = request->gid();
+    uint32_t mode = request->mode();
+
+    MetaStatusCode status = inodeManager_->CreateRootInode(fsId, uid, gid,
+                                                            mode);
+    response->set_statuscode(status);
+    if (status != MetaStatusCode::OK) {
+        LOG(ERROR) << "CreateRootInode fail, fsId = " << fsId
+                   << ", uid = " << uid
+                   << ", gid = " << gid
+                   << ", mode = " << mode
+                   << ", retCode = " << MetaStatusCode_Name(status);
+    }
+    return;
+}
+
+void MetaServerServiceImpl::UpdateInode(
+                    ::google::protobuf::RpcController* controller,
+                    const ::curvefs::metaserver::UpdateInodeRequest* request,
+                    ::curvefs::metaserver::UpdateInodeResponse* response,
+                    ::google::protobuf::Closure* done) {
+    brpc::ClosureGuard doneGuard(done);
+    brpc::Controller* cntl = static_cast<brpc::Controller*>(controller);
+    uint32_t fsId = request->fsid();
+    uint64_t inodeId = request->inodeid();
+    if (request->has_volumeextentlist() && request->has_s3chunkinfolist()) {
+        LOG(ERROR) << "only one of type space info, choose volume or s3";
+        response->set_statuscode(MetaStatusCode::PARAM_ERROR);
+        return;
+    }
+
+    Inode inode;
+    MetaStatusCode status = inodeManager_->GetInode(fsId, inodeId, &inode);
+    if (status != MetaStatusCode::OK) {
+        response->set_statuscode(status);
+        return;
+    }
+
+    bool needUpdate = false;
+
+    #define UPDATE_INODE(param)  \
```

Comment:
Maybe it's better to define it in header file.

Reply:
It only use here. If in the future othre place use this code, it can move to header file.